### PR TITLE
[Reader Mode] Add theme support features from `theme.json`

### LINKED ISSUE #7471: Invisible cover block background colors
### Bug Description
Cover block background colors do not work as expected.

### Expected Behaviour
Should render as in the editor.

### Screenshots
Screenshot at Feb 23 17-37-57
Screenshot at Feb 23 17-37-57

### PHP Version
8.1

### Plugin Version
2.4.0

### AMP plugin template mode
Reader

### WordPress Version
6.1.1

### Site Health
No issues

### Gutenberg Version
No response

### OS(s) Affected
No response

### Browser(s) Affected
No response

### Device(s) Affected
No response

### Acceptance Criteria
No response

### Implementation Brief
No response

### QA Testing Instructions
No response

### Demo
No response

### Changelog Entry
No response

### PR DESCRIPTION
**Pull request by thelovekesh**
## Summary

Implemented in #6042, reader themes provide support for producing color palettes, gradient presets, and font size styles. This feature leverages `get theme support` to identify the necessary presets for generating the styles. 

Although `theme.json` can be used by any theme, it serves as a foundation for the new block themes. This PR seeks to enable theme support features styles generation using the `theme.json` presets to address the block styles breaking issues in reader mode.

Fixes #7471

## Checklist

- [x] My code is tested and passes existing [tests](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines#tests).
- [x] My code follows the [Engineering Guidelines](https://github.com/ampproject/amp-wp/wiki/Engineering-Guidelines) (updates are often made to the guidelines, check it out periodically).

### CODE REVIEW

**Comment by thelovekesh** [comment]
@westonruter Although the support for `theme.json` was added in WP 5.8, the required APIs like `wp_get_global_settings` was added in WP 5.9. Should I look for a workaround to get it worked with WP 5.8 as well or we are good to support it on WP 5.9+?

---
1. Right now our CSS unit measurement doesn't look robust as it only handles all values in `px`

https://github.com/ampproject/amp-wp/blob/2a3d437a50290c93359f50c54c56f55b87313df6/src/ReaderThemeSupportFeatures.php#L421

but there might be cases when someone defines them in `em` or `rem`.

2. Also, we will need to calculate values for `clamp()` to give support for fluid typography. 

To solve the above scenarios WP 6.1 has two functions:
- [`wp_get_typography_value_and_unit()`](https://github.com/WordPress/WordPress/blob/9caf1c4adeddff2577c24d622ebbbf278a671271/wp-includes/block-supports/typography.php#L297)
- [`wp_get_computed_fluid_typography_value()`](https://github.com/WordPress/WordPress/blob/9caf1c4adeddff2577c24d622ebbbf278a671271/wp-includes/block-supports/typography.php#L385)

but the caveat is they are available prior to WP 6.1.  

I'm considering defining them in our own class right now. Please let me know your thoughts?

**Comment by westonruter** [comment]
If someone is using block themes, I'd say it's highly unlikely that they would be holding off on updating WordPress core. Therefore, if you want to require WordPress 6.1+ to fully support fluid typography, that's fine with me.

**Comment by thelovekesh** [comment]
But I think we should update the way of determining measurement units. Someone can use different measurement unit even when fluid typography is not being used.

File: src/ReaderThemeSupportFeatures.php
Comment (thelovekesh) [inline src/ReaderThemeSupportFeatures.php:641]: We can ignore coverage for this copied method?

**Comment by thelovekesh** [comment]
PR can be tested for fluid typography by disabling CSS tree-shaking till #7291 get fixed.

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php:641]: Sure

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: Minor point, but I don't think `esc_attr()` is right here. I think `strip_tags()` is the right choice. See https://github.com/WordPress/wordpress-develop/blob/200868214a1ae0a108dac491677ba82e7541fc8d/src/wp-includes/theme.php#L1891-L1896

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php:650]: Shouldn't this actually be the version that is in core?

Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php:650]: Well, it's not in core. Nevermind.

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: This would seem simpler as:

```suggestion
		return is_readable( get_stylesheet_directory() . '/theme.json' ) || is_readable( get_template_directory() . '/theme.json' );
```

Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php]: Actually, I suggest using `wp_theme_has_theme_json()` if it exists, or else copy the function as a method into this class:

https://github.com/WordPress/wordpress-develop/blob/200868214a1ae0a108dac491677ba82e7541fc8d/src/wp-includes/global-styles-and-settings.php#L384-L414

Note how it caches the file exists check.

File: src/ReaderThemeSupportFeatures.php
Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php]: It does make sense, but I am not sure if we will ever get tags here. Or better to never assume and update it to use `strip_tags()` 🤔

File: src/ReaderThemeSupportFeatures.php
Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php]: > Note how it caches the file exists check.

Oh yes. Bette to copy it since we will need to check for theme.json on WP 5.9+ and this function is added in WP 6.2.

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php]: Better to never assume, yeah

Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php]: Just realized, we will still need the escaping, as `strip_tags` is not an escaping utility.

Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php]: Well, it is somewhat. It's what core uses in `wp_custom_css_cb()`.

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php:293]: There is missing test coverage in this method. Is that feasible to add?

Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php:293]: Tests are there but they were wrongfully skipped due to https://github.com/ampproject/amp-wp/pull/7481/commits/f1d16af946f18b3f2e9d44fb42e7724be265ffa3.

I have fixed it in https://github.com/ampproject/amp-wp/pull/7481/commits/722476c6b1c61b26b08eda42d954a95e115398c5

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: This can just use `strip_tags()` since `esc_attr()`  could introduce entities which break CSS.

```suggestion
					? strip_tags( wp_get_typography_font_size_value( $font_size ) )
					: strip_tags( $font_size[ self::KEY_SIZE ] )
```

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: This should defer to `wp_theme_has_theme_json()` if it exists:

```suggestion
		if ( function_exists( 'wp_theme_has_theme_json' ) ) {
			return wp_theme_has_theme_json();
		}
		// wp_get_global_settings() is only available in WP 5.9+.
```

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php:646]:
```suggestion
		if ( function_exists( 'wp_get_typography_value_and_unit' ) ) {
			return wp_get_typography_value_and_unit( $raw_value, $options );
		}
		if ( ! is_string( $raw_value ) && ! is_int( $raw_value ) && ! is_float( $raw_value ) ) {
```

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php:307]: I confess I am not up to speed on `theme.json`, but wouldn't the `theme` value supercede the `default` value? In other words, shouldn't these two conditions be swapped? (I'd be grateful to be educated otherwise!)

Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php:307]: The following is the output (in array) when we use `wp_get_global_settings` to retrieve theme features from `theme.json`:

Here, `theme` is derived from `theme.json`, and the `default` preset is coming from the block editor.

<details><summary>Theme Features</summary>

```json
{
  "default": [
    {
      "name": "Black",
      "slug": "black",
      "color": "#000000"
    },
    {
      "name": "Cyan bluish gray",
      "slug": "cyan-bluish-gray",
      "color": "#abb8c3"
    },
    {
      "name": "White",
      "slug": "white",
      "color": "#ffffff"
    },
    {
      "name": "Pale pink",
      "slug": "pale-pink",
      "color": "#f78da7"
    },
    {
      "name": "Vivid red",
      "slug": "vivid-red",
      "color": "#cf2e2e"
    },
    {
      "name": "Luminous vivid orange",
      "slug": "luminous-vivid-orange",
      "color": "#ff6900"
    },
    {
      "name": "Luminous vivid amber",
      "slug": "luminous-vivid-amber",
      "color": "#fcb900"
    },
    {
      "name": "Light green cyan",
      "slug": "light-green-cyan",
      "color": "#7bdcb5"
    },
    {
      "name": "Vivid green cyan",
      "slug": "vivid-green-cyan",
      "color": "#00d084"
    },
    {
      "name": "Pale cyan blue",
      "slug": "pale-cyan-blue",
      "color": "#8ed1fc"
    },
    {
      "name": "Vivid cyan blue",
      "slug": "vivid-cyan-blue",
      "color": "#0693e3"
    },
    {
      "name": "Vivid purple",
      "slug": "vivid-purple",
      "color": "#9b51e0"
    }
  ],
  "theme": [
    {
      "color": "#ffffff",
      "name": "Base",
      "slug": "base"
    },
    {
      "color": "#000000",
      "name": "Contrast",
      "slug": "contrast"
    },
    {
      "color": "#9DFF20",
      "name": "Primary",
      "slug": "primary"
    },
    {
      "color": "#345C00",
      "name": "Secondary",
      "slug": "secondary"
    },
    {
      "color": "#F6F6F6",
      "name": "Tertiary",
      "slug": "tertiary"
    }
  ]
}
```

</details> 

<details><summary>Block Editor View</summary>
<p>

image

</p>
</details> 

Therefore I believe they do not intend for one to supersede the other. The user may select from the `theme` preset or `default` preset at will.

Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php:307]: Ah, it is not merging associative arrays, so my concern is invalid.

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: While the `theme_has_theme_json` method has a check for `wp_get_global_settings`, it also now short-circuits if the `wp_theme_has_theme_json` function exists. Therefore, I think it may be safer just to move the function exists check here:

```suggestion
			if ( $this->theme_has_theme_json() && function_exists( 'wp_get_global_settings' ) ) {
				$feature_value   = [];
				$global_settings = wp_get_global_settings( self::SUPPORTED_THEME_JSON_FEATURES[ $feature_key ], 'theme' );
```

File: src/ReaderThemeSupportFeatures.php
Comment (westonruter) [inline src/ReaderThemeSupportFeatures.php]: See above.

```suggestion
```

Comment (westonruter) [reply src/ReaderThemeSupportFeatures.php]: This then makes the function an actual 1:1 copy of `wp_theme_has_theme_json()`

File: tests/php/src/ReaderThemeSupportFeaturesTest.php
Comment (westonruter) [inline tests/php/src/ReaderThemeSupportFeaturesTest.php]: This shouldn't be necessary since the test tear-down handles this automatically.

```suggestion
```

File: tests/php/src/ReaderThemeSupportFeaturesTest.php
Comment (westonruter) [inline tests/php/src/ReaderThemeSupportFeaturesTest.php]:
```suggestion
```

File: src/ReaderThemeSupportFeatures.php
Comment (thelovekesh) [reply src/ReaderThemeSupportFeatures.php]: Yes, it seems much better.

`wp_theme_has_theme_json` is introduced in WP 6.2 so `wp_get_global_settings` will exists, but this solution looks cleaner to me.

**Comment by westonruter** [comment]
This seems very close. I'm testing with 2023 with the following content:

```html
<!-- wp:paragraph {"fontSize":"xx-large"} -->
<p class="has-xx-large-font-size">Hello!</p>
<!-- /wp:paragraph -->
```

Primary Theme (2023) | Reader Theme (2021) Before | Reader Theme (2021) After
---------------------|-----------------------------|--------------------------
image | image | image

So that's great. What's not quite there yet is the Legacy theme:

> image

Something's not right with margin or line-height or something.

**Comment by thelovekesh** [comment]
Seems like on legacy themes `line-height` is causing issues.

<details><summary>With <code>line-height: 1.75em</code></summary>

image
</details> 

<details><summary>With <code>line-height: 1.5</code></summary>

image
</details>

**Comment by westonruter** [comment]
In TT3, there is this rule applying to the content container:

```css
body .is-layout-flow > * + * {
    margin-block-start: 1.5rem;
    margin-block-end: 0;
}
```

Applying these styles to `.amp-wp-article-content` mostly fixes the issue:

image

Not perfectly, but better.

**Comment by westonruter** [comment]
OK, let's change this:

https://github.com/ampproject/amp-wp/blob/4337065a80df965a4d49d2858e6fcced339b90bc/templates/style.php#L116

To just `1.75`.

**Comment by westonruter** [comment]
`line-height: 1.75em` | `line-height: 1.75`
---------|--------
image | image

**Comment by westonruter** [comment]
Something else that isn't quite right, even with `line-height:1.75`. Adding a background color to a paragraph that has XXL font size:

```html
<!-- wp:paragraph {"backgroundColor":"primary","fontSize":"xx-large"} -->
<p class="has-primary-background-color has-background has-xx-large-font-size">Hello!</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent interdum, elit non cursus tincidunt, nibh nibh congue metus, ac gravida nunc ipsum ut augue. Donec tempor pulvinar maximus. Sed et tellus nec sapien vulputate sollicitudin. Nam auctor nulla imperdiet, sodales leo ultricies, cursus metus. Aenean ullamcorper dui vel neque luctus faucibus non vel odio. Aenean convallis mattis nisi, nec fringilla sem sagittis eget. Duis ut quam venenatis, blandit elit id, iaculis ipsum. Curabitur tristique euismod nisl, a finibus massa dignissim ac. Nullam nisl orci, imperdiet at libero sed, pulvinar vehicula lacus. Duis eleifend sollicitudin pharetra. Maecenas sapien nunc, mollis a velit sed, ultricies accumsan tortor. Ut accumsan sapien id turpis eleifend, at eleifend orci volutpat. Etiam massa arcu, congue sit amet sem vitae, gravida efficitur sem. Curabitur ornare velit velit, ut accumsan tellus convallis at. Nunc sagittis nisl mauris, nec hendrerit leo sagittis et.</p>
<!-- /wp:paragraph -->
```

TT3 | TT2 Reader | Legacy Reader
-----|-----------|--------------
image | image | image

**Comment by thelovekesh** [comment]
Seems like the `xx-large` font size requires more width on the screen but we have restricted it to `840px` in `.amp-wp-article` which wraps the whole article in it.

https://github.com/ampproject/amp-wp/blob/d03e17921daab6f614d463d8274f8154a6f20972/templates/style.php#L203

**Comment by westonruter** [comment]
Oh, if I actually look at Legacy in a mobile viewport it renders fine:

[Screen recording 2023-03-17 13.20.20.webm](https://user-images.githubusercontent.com/134745/226028786-3d3cec77-d710-4a4e-92a2-0c15c5789ce1.webm)

We could fix this by adding negative margins:

```css
p.has-background {
    margin-left: -2.375em;
    margin-right: -2.375em;
}
```

[Screen recording 2023-03-17 13.24.26.webm](https://user-images.githubusercontent.com/134745/226030205-1257d667-ed52-40a3-a17e-0474ccf8bf02.webm)

**Comment by westonruter** [comment]
Another case to check for: groups with backgrounds.

```html
<!-- wp:group {"style":{"spacing":{"padding":{"top":"var:preset|spacing|40","right":"var:preset|spacing|40","bottom":"var:preset|spacing|40","left":"var:preset|spacing|40"}}},"backgroundColor":"primary","layout":{"type":"constrained"}} -->
<div class="wp-block-group has-primary-background-color has-background" style="padding-top:var(--wp--preset--spacing--40);padding-right:var(--wp--preset--spacing--40);padding-bottom:var(--wp--preset--spacing--40);padding-left:var(--wp--preset--spacing--40)"><!-- wp:paragraph {"backgroundColor":"luminous-vivid-amber"} -->
<p class="has-luminous-vivid-amber-background-color has-background">First</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph {"backgroundColor":"vivid-red"} -->
<p class="has-vivid-red-background-color has-background">Second</p>
<!-- /wp:paragraph --></div>
<!-- /wp:group -->
```

Should look like this:

image

File: .github/workflows/build-test-measure.yml
Comment (westonruter) [inline .github/workflows/build-test-measure.yml:580]: Why specifically?

Comment (thelovekesh) [reply .github/workflows/build-test-measure.yml:580]: Sometimes external-http tests may fail the workflow, causing main workflow to exit early.

**Comment by westonruter** [comment]
Looking good!

Non-AMP | Desktop Legacy AMP | Mobile Legacy AMP
--|--|--
image | image | image

**Review by westonruter — APPROVED** [review]

**Pull request merged**
